File: src/integrations/graph.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

/**
 * Microsoft Graph — SharePoint and Teams search.
 *
 * Extends the Graph client to cover the two surfaces that hold the other
 * half of a law firm's institutional knowledge:
 *
 *   SharePoint — document libraries, matter workspaces, precedent repositories.
 *   Teams      — channel conversations, meeting notes, matter-specific team chats.
 *
 * Used by:
 *   • Briefing swarm spokes (pull intel into the chalkboard)
 *   • Channel bot (post findings back to Teams channels)
 *   • Precedent search (find documents in SharePoint matter sites)
 *
 * Auth: reuses the GRAPH_* credentials already configured for email.
 * The same app registration needs additional API permissions:
 *   SharePoint — Sites.Read.All (application)
 *   Teams      — ChannelMessage.Read.All (application)
 *
 * All HTTP calls: 15 s timeout, 512 KB response cap, no credentials in logs.
 */

import { Config } from "../config.js";
import { logger } from "../logger.js";

const GRAPH_BASE = "https://graph.microsoft.com/v1.0";
const REQUEST_TIMEOUT_MS = 15_000;
const MAX_RESPONSE_BYTES = 512 * 1024;

// ─── Token cache (shared with email/client.ts via re-export) ─────────────────

let _cachedToken: { token: string; expiresAt: number } | null = null;
let _graphTokenPromise: Promise<string> | null = null;

export async function getGraphToken(): Promise<string> {
  const cfg = Config.email.graph;
  if (!cfg.enabled) throw new Error("Microsoft Graph not configured");

  if (cfg.accessToken) return cfg.accessToken;
  if (_cachedToken && Date.now() < _cachedToken.expiresAt - 60_000) return _cachedToken.token;

  // In-flight deduplication: only one token request at a time
  _graphTokenPromise ??= (async () => {
    const body = new URLSearchParams({
      client_id: cfg.clientId, client_secret: cfg.clientSecret,
      scope: "https://graph.microsoft.com/.default",
      grant_type: "client_credentials",
    });
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), REQUEST_TIMEOUT_MS);
    try {
      const res = await fetch(
        `https://login.microsoftonline.com/${cfg.tenantId}/oauth2/v2.0/token`,
        { method: "POST", headers: { "content-type": "application/x-www-form-urlencoded" }, body, signal: ctrl.signal },
      );
      clearTimeout(timer);
      if (!res.ok) throw new Error(`Graph token failed: ${res.status}`);
      const data = (await res.json()) as { access_token: string; expires_in: number };
      _cachedToken = { token: data.access_token, expiresAt: Date.now() + data.expires_in * 1000 };
      return data.access_token;
    } catch (err) { clearTimeout(timer); throw err; }
  })().finally(() => { _graphTokenPromise = null; });
  return _graphTokenPromise;
}

// ─── URL validation ───────────────────────────────────────────────────────────

/**
 * Validate that an absolute URL returned from a Graph @odata.nextLink is
 * actually a Graph URL. Prevents SSRF if a malicious response forges the link.
 */
export function validateGraphUrl(url: string): string {
  const parsed = new URL(url);
  if (parsed.hostname !== "graph.microsoft.com") {
    throw new Error(`Blocked non-Graph URL: ${parsed.hostname}`);
  }
  return url;
}

// ─── Shared fetch ─────────────────────────────────────────────────────────────

export async function graphFetch(path: string, token: string): Promise<unknown> {
  // Always build relative to GRAPH_BASE. For @odata.nextLink pagination, callers
  // must pass the result of validateGraphUrl() to prevent SSRF.
  const url = `${GRAPH_BASE}${path}`;
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), REQUEST_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      headers: { authorization: `Bearer ${token}`, accept: "application/json", "Content-Type": "application/json" },
      signal: ctrl.signal,
    });
    clearTimeout(timer);
    if (!res.ok) throw new Error(`Graph API error ${res.status}: ${path}`);
    const reader = res.body?.getReader();
    if (!reader) return res.json();
    let bytes = 0; const chunks: Uint8Array[] = [];
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      bytes += value.length;
      if (bytes > MAX_RESPONSE_BYTES) throw new Error("Graph response too large");
      chunks.push(value);
    }
    const merged = chunks.reduce((a, c) => { const n = new Uint8Array(a.length + c.length); n.set(a); n.set(c, a.length); return n; }, new Uint8Array(0));
    return JSON.parse(new TextDecoder().decode(merged));
  } catch (err) { clearTimeout(timer); throw err; }
}

export async function graphPost(path: string, token: string, body: unknown): Promise<unknown> {
  const url = `${GRAPH_BASE}${path}`;
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), REQUEST_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { authorization: `Bearer ${token}`, "content-type": "application/json", accept: "application/json" },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    clearTimeout(timer);
    if (!res.ok) {
      const text = await res.text().catch(() => "");
      throw new Error(`Graph POST error ${res.status}: ${path} — ${text.slice(0, 200)}`);
    }
    const text = await res.text();
    return text ? JSON.parse(text) : null;
  } catch (err) { clearTimeout(timer); throw err; }
}

// ─── SharePoint ───────────────────────────────────────────────────────────────

export interface SharePointFile {
  id: string;
  name: string;
  webUrl: string;
  lastModified: string;
  size: number;
  siteId: string;
  siteName: string;
  matterRef?: string;
}

/**
 * Search across all SharePoint sites for files matching a query.
 * Uses Graph unified search (entityTypes: driveItem) — covers all sites
 * the app has Sites.Read.All access to.
 */
export async function searchSharePoint(
  query: string,
  opts: { maxResults?: number; daysBack?: number } = {},
): Promise<SharePointFile[]> {
  if (!Config.email.graph.enabled) return [];
  const token = await getGraphToken();
  const max = opts.maxResults ?? 20;

  try {
    const body = {
      requests: [{
        entityTypes: ["driveItem"],
        query: { queryString: query },
        from: 0,
        size: max,
        fields: ["id", "name", "webUrl", "lastModifiedDateTime", "size", "parentReference"],
      }],
    };
    const result = await graphPost("/search/query", token, body) as Record<string, unknown>;
    const hits = (result["value"] as Array<Record<string, unknown>>)?.[0]
      ?.["hitsContainers"] as Array<Record<string, unknown>>;
    if (!Array.isArray(hits)) return [];

    return hits.flatMap((hc) => {
      const items = hc["hits"] as Array<Record<string, unknown>>;
      return Array.isArray(items) ? items.map((h): SharePointFile => {
        const r = h["resource"] as Record<string, unknown>;
        const parent = r["parentReference"] as Record<string, unknown>;
        return {
          id: String(r["id"] ?? ""),
          name: String(r["name"] ?? ""),
          webUrl: String(r["webUrl"] ?? ""),
          lastModified: String(r["lastModifiedDateTime"] ?? ""),
          size: Number(r["size"] ?? 0),
          siteId: String(parent?.["siteId"] ?? ""),
          siteName: String(parent?.["siteName"] ?? ""),
          matterRef: extractMatterRef(String(r["name"] ?? "")),
        };
      }) : [];
    }).slice(0, max);
  } catch (err) {
    logger.warn("SharePoint search failed", { error: (err as Error).message });
    return [];
  }
}

// ─── Teams ────────────────────────────────────────────────────────────────────

export interface TeamsMessage {
  id: string;
  teamName: string;
  channelName: string;
  channelId: string;
  from: string;
  createdAt: string;
  body: string;
  webUrl?: string;
  matterRef?: string;
}

/**
 * Search Teams channel messages across all teams the app can see.
 * Uses Graph unified search (entityTypes: chatMessage).
 */
export async function searchTeamsMessages(
  query: string,
  opts: { maxResults?: number; daysBack?: number } = {},
): Promise<TeamsMessage[]> {
  if (!Config.email.graph.enabled) return [];
  const token = await getGraphToken();
  const max = opts.maxResults ?? 20;

  try {
    const body = {
      requests: [{
        entityTypes: ["chatMessage"],
        query: { queryString: query },
        from: 0,
        size: max,
      }],
    };
    const result = await graphPost("/search/query", token, body) as Record<string, unknown>;
    const hits = (result["value"] as Array<Record<string, unknown>>)?.[0]
      ?.["hitsContainers"] as Array<Record<string, unknown>>;
    if (!Array.isArray(hits)) return [];

    return hits.flatMap((hc) => {
      const items = hc["hits"] as Array<Record<string, unknown>>;
      return Array.isArray(items) ? items.map((h): TeamsMessage => {
        const r = h["resource"] as Record<string, unknown>;
        const from = r["from"] as Record<string, unknown>;
        const fromName = String(
          (from?.["user"] as Record<string, unknown>)?.["displayName"]
          ?? (from?.["application"] as Record<string, unknown>)?.["displayName"]
          ?? "Unknown",
        );
        const bodyText = String((r["body"] as Record<string, unknown>)?.["content"] ?? "").replace(/<[^>]+>/g, " ").slice(0, 400);
        return {
          id: String(r["id"] ?? ""),
          teamName: String((r["channelIdentity"] as Record<string, unknown>)?.["teamId"] ?? ""),
          channelName: String((r["channelIdentity"] as Record<string, unknown>)?.["channelId"] ?? ""),
          channelId: String((r["channelIdentity"] as Record<string, unknown>)?.["channelId"] ?? ""),
          from: fromName,
          createdAt: String(r["createdDateTime"] ?? ""),
          body: bodyText,
          webUrl: String(r["webUrl"] ?? "") || undefined,
          matterRef: extractMatterRef(bodyText),
        };
      }) : [];
    }).slice(0, max);
  } catch (err) {
    logger.warn("Teams search failed", { error: (err as Error).message });
    return [];
  }
}

/**
 * Post a message to a Teams channel via the Graph API.
 * Requires ChannelMessage.Send (delegated) or use incoming webhook URL.
 */
export async function postToTeamsChannel(
  channelId: string,
  teamId: string,
  content: string,
): Promise<void> {
  const token = await getGraphToken();
  await graphPost(
    `/teams/${encodeURIComponent(teamId)}/channels/${encodeURIComponent(channelId)}/messages`,
    token,
    { body: { contentType: "html", content } },
  );
}

/**
 * Post to a Teams channel via an Incoming Webhook URL (simpler, no app auth needed).
 * TEAMS_INCOMING_WEBHOOK_URL must be set.
 */
export async function postToTeamsWebhook(
  webhookUrl: string,
  title: string,
  text: string,
  facts?: Array<{ name: string; value: string }>,
): Promise<void> {
  // Validate URL is a genuine Teams webhook before sending any data
  try {
    const u = new URL(webhookUrl);
    if (!["outlook.office.com", "outlook.office365.com"].some(
      (host) => u.hostname === host || u.hostname.endsWith(".webhook.office.com"),
    )) {
      throw new Error(`Not a valid Teams webhook hostname: ${u.hostname}`);
    }
    if (u.protocol !== "https:") throw new Error("Teams webhook URL must use HTTPS");
  } catch (err) {
    logger.warn("Blocked invalid Teams webhook URL", { error: (err as Error).message });
    throw new Error(`Invalid Teams webhook URL: ${(err as Error).message}`);
  }

  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), REQUEST_TIMEOUT_MS);
  const card = {
    "@type": "MessageCard",
    "@context": "https://schema.org/extensions",
    themeColor: "0076D7",
    summary: title,
    sections: [{
      activityTitle: title,
      activityText: text,
      facts: facts ?? [],
    }],
  };
  try {
    const res = await fetch(webhookUrl, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(card),
      signal: ctrl.signal,
    });
    clearTimeout(timer);
    if (!res.ok) throw new Error(`Teams webhook POST failed: ${res.status}`);
  } catch (err) { clearTimeout(timer); throw err; }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function extractMatterRef(text: string): string | undefined {
  const patterns = [/\[([A-Z]{1,5}[-/]\d{3,8})\]/, /\b([A-Z]{1,5}[-/]\d{4,8})\b/];
  for (const p of patterns) {
    const m = text.match(p);
    if (m?.[1]) return m[1];
  }
  return undefined;
}

File: biglaw-go/internal/integrations/graph.go
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

// Microsoft Graph — SharePoint search, Teams message search, webhook posting.
// Auth: client-credentials OAuth2 using GRAPH_TENANT_ID / CLIENT_ID / CLIENT_SECRET.
// All calls: 15 s timeout, 512 KB response cap. Credentials never logged.

package integrations

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"sync"
	"time"

	"github.com/discover-legal/biglaw-go/internal/strutil"
)

const (
	graphBase     = "https://graph.microsoft.com/v1.0"
	graphTimeout  = 15 * time.Second
	graphMaxBytes = 512 * 1024
)

// ─── Token cache ──────────────────────────────────────────────────────────────

var (
	graphTokenMu  sync.Mutex
	graphTokenVal string
	graphTokenExp time.Time
)

// GetGraphToken returns a valid MS Graph bearer token.
// Returns "" if not configured.
func GetGraphToken() (string, error) {
	// Pre-obtained token (dev mode)
	if t := os.Getenv("GRAPH_ACCESS_TOKEN"); t != "" {
		return t, nil
	}
	tenantID := os.Getenv("GRAPH_TENANT_ID")
	clientID := os.Getenv("GRAPH_CLIENT_ID")
	clientSecret := os.Getenv("GRAPH_CLIENT_SECRET")
	if tenantID == "" || clientID == "" || clientSecret == "" {
		return "", fmt.Errorf("Microsoft Graph not configured")
	}

	graphTokenMu.Lock()
	defer graphTokenMu.Unlock()
	if graphTokenVal != "" && time.Now().Before(graphTokenExp.Add(-60*time.Second)) {
		return graphTokenVal, nil
	}

	form := url.Values{
		"client_id":     {clientID},
		"client_secret": {clientSecret},
		"scope":         {"https://graph.microsoft.com/.default"},
		"grant_type":    {"client_credentials"},
	}
	tokenURL := fmt.Sprintf("https://login.microsoftonline.com/%s/oauth2/v2.0/token", tenantID)
	resp, err := httpPost(tokenURL, "application/x-www-form-urlencoded", bytes.NewBufferString(form.Encode()), "")
	if err != nil {
		return "", err
	}
	var tok struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int    `json:"expires_in"`
	}
	if err := json.Unmarshal(resp, &tok); err != nil {
		return "", err
	}
	graphTokenVal = tok.AccessToken
	graphTokenExp = time.Now().Add(time.Duration(tok.ExpiresIn) * time.Second)
	return tok.AccessToken, nil
}

// ─── HTTP helpers ─────────────────────────────────────────────────────────────

func httpGet(urlStr, bearer string) ([]byte, error) {
	client := &http.Client{Timeout: graphTimeout}
	req, err := http.NewRequest("GET", urlStr, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	if bearer != "" {
		req.Header.Set("Authorization", "Bearer "+bearer)
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("Graph GET %s: HTTP %d", urlStr, resp.StatusCode)
	}
	return io.ReadAll(io.LimitReader(resp.Body, graphMaxBytes))
}

func httpPost(urlStr, contentType string, body io.Reader, bearer string) ([]byte, error) {
	client := &http.Client{Timeout: graphTimeout}
	req, err := http.NewRequest("POST", urlStr, body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", contentType)
	req.Header.Set("Accept", "application/json")
	if bearer != "" {
		req.Header.Set("Authorization", "Bearer "+bearer)
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return nil, fmt.Errorf("Graph POST %s: HTTP %d — %s", urlStr, resp.StatusCode, b)
	}
	return io.ReadAll(io.LimitReader(resp.Body, graphMaxBytes))
}

func graphPost(path, token string, payload interface{}) ([]byte, error) {
	urlStr := path
	if len(path) > 0 && path[0] == '/' {
		urlStr = graphBase + path
	}
	b, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	return httpPost(urlStr, "application/json", bytes.NewReader(b), token)
}

// ─── SharePoint ───────────────────────────────────────────────────────────────

type SharePointFile struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	WebURL       string `json:"webUrl"`
	LastModified string `json:"lastModified"`
	Size         int64  `json:"size"`
	SiteID       string `json:"siteId"`
	SiteName     string `json:"siteName"`
	MatterRef    string `json:"matterRef,omitempty"`
}

func SearchSharePoint(query string, maxResults int) ([]SharePointFile, error) {
	token, err := GetGraphToken()
	if err != nil {
		return nil, nil // not configured — return empty silently
	}
	if maxResults <= 0 {
		maxResults = 20
	}
	body := map[string]interface{}{
		"requests": []map[string]interface{}{{
			"entityTypes": []string{"driveItem"},
			"query":       map[string]string{"queryString": query},
			"from":        0,
			"size":        maxResults,
			"fields":      []string{"id", "name", "webUrl", "lastModifiedDateTime", "size", "parentReference"},
		}},
	}
	data, err := graphPost("/search/query", token, body)
	if err != nil {
		return nil, err
	}
	var result struct {
		Value []struct {
			HitsContainers []struct {
				Hits []struct {
					Resource map[string]interface{} `json:"resource"`
				} `json:"hits"`
			} `json:"hitsContainers"`
		} `json:"value"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	var files []SharePointFile
	for _, v := range result.Value {
		for _, hc := range v.HitsContainers {
			for _, h := range hc.Hits {
				r := h.Resource
				parent, _ := r["parentReference"].(map[string]interface{})
				f := SharePointFile{
					ID:           str(r["id"]),
					Name:         str(r["name"]),
					WebURL:       str(r["webUrl"]),
					LastModified: str(r["lastModifiedDateTime"]),
					SiteID:       str(parent["siteId"]),
					SiteName:     str(parent["siteName"]),
				}
				if s, ok := r["size"].(float64); ok {
					f.Size = int64(s)
				}
				f.MatterRef = extractMatterRef(f.Name)
				files = append(files, f)
			}
		}
	}
	if len(files) > maxResults {
		files = files[:maxResults]
	}
	return files, nil
}

// ─── Teams messages ───────────────────────────────────────────────────────────

type TeamsMessage struct {
	ID          string `json:"id"`
	TeamName    string `json:"teamName"`
	ChannelName string `json:"channelName"`
	ChannelID   string `json:"channelId"`
	From        string `json:"from"`
	CreatedAt   string `json:"createdAt"`
	Body        string `json:"body"`
	WebURL      string `json:"webUrl,omitempty"`
	MatterRef   string `json:"matterRef,omitempty"`
}

func SearchTeamsMessages(query string, maxResults int) ([]TeamsMessage, error) {
	token, err := GetGraphToken()
	if err != nil {
		return nil, nil
	}
	if maxResults <= 0 {
		maxResults = 20
	}
	body := map[string]interface{}{
		"requests": []map[string]interface{}{{
			"entityTypes": []string{"chatMessage"},
			"query":       map[string]string{"queryString": query},
			"from":        0,
			"size":        maxResults,
		}},
	}
	data, err := graphPost("/search/query", token, body)
	if err != nil {
		return nil, err
	}
	var result struct {
		Value []struct {
			HitsContainers []struct {
				Hits []struct {
					Resource map[string]interface{} `json:"resource"`
				} `json:"hits"`
			} `json:"hitsContainers"`
		} `json:"value"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	var msgs []TeamsMessage
	for _, v := range result.Value {
		for _, hc := range v.HitsContainers {
			for _, h := range hc.Hits {
				r := h.Resource
				chId := r["channelIdentity"]
				var chMap map[string]interface{}
				if m, ok := chId.(map[string]interface{}); ok {
					chMap = m
				}
				fromObj, _ := r["from"].(map[string]interface{})
				fromName := ""
				if u, ok := fromObj["user"].(map[string]interface{}); ok {
					fromName = str(u["displayName"])
				} else if a, ok := fromObj["application"].(map[string]interface{}); ok {
					fromName = str(a["displayName"])
				}
				bodyObj, _ := r["body"].(map[string]interface{})
				bodyText := stripHTML(str(bodyObj["content"]))
				if len(bodyText) > 400 {
					bodyText = strutil.Truncate(bodyText, 400)
				}
				m := TeamsMessage{
					ID:        str(r["id"]),
					ChannelID: str(chMap["channelId"]),
					From:      fromName,
					CreatedAt: str(r["createdDateTime"]),
					Body:      bodyText,
					WebURL:    str(r["webUrl"]),
				}
				m.MatterRef = extractMatterRef(bodyText)
				msgs = append(msgs, m)
			}
		}
	}
	if len(msgs) > maxResults {
		msgs = msgs[:maxResults]
	}
	return msgs, nil
}

// ─── Teams Incoming Webhook ───────────────────────────────────────────────────

type WebhookFact struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

// PostToTeamsWebhook posts a MessageCard to a Teams Incoming Webhook URL.
func PostToTeamsWebhook(webhookURL, title, text string, facts []WebhookFact) error {
	card := map[string]interface{}{
		"@type":      "MessageCard",
		"@context":   "http://schema.org/extensions",
		"themeColor": "0076D7",
		"summary":    title,
		"sections": []map[string]interface{}{{
			"activityTitle": title,
			"activityText":  text,
			"facts":         facts,
		}},
	}
	b, err := json.Marshal(card)
	if err != nil {
		return err
	}
	_, err = httpPost(webhookURL, "application/json", bytes.NewReader(b), "")
	return err
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

var (
	matterRefRe1 = regexp.MustCompile(`\[([A-Z]{1,5}[-/]\d{3,8})\]`)
	matterRefRe2 = regexp.MustCompile(`\b([A-Z]{1,5}[-/]\d{4,8})\b`)
	htmlTagRe    = regexp.MustCompile(`<[^>]+>`)
)

func extractMatterRef(s string) string {
	if m := matterRefRe1.FindStringSubmatch(s); len(m) > 1 {
		return m[1]
	}
	if m := matterRefRe2.FindStringSubmatch(s); len(m) > 1 {
		return m[1]
	}
	return ""
}

func stripHTML(s string) string {
	return htmlTagRe.ReplaceAllString(s, " ")
}

func str(v interface{}) string {
	if v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%v", v)
}
